std/range/package.d

```diff
@@ -6767,6 +6767,543 @@ if (isForwardRange!Source && hasLength!Source)
+     * possible. Further, using assumeNonEmpty formalizes the assumption of
+     * non-null.
+     * Returns: The held value
+     * Throws: `AssertError` if this instance is empty
+     * Note: `opt.assumeNonEmpty = null;` does not empty the instance; to empty
+     * the instance, assign to $(LREF none).
+     * Bugs: This function is currently not `nothrow` or `@nogc`.
+     * $(LREF Option.front) can be used as a workaround when these attributes
+     * are desired. See $(BUGZILLA 12647).
+     */
+    // this function is not nothrow due to bug 12647, and not @nogc.
+    // onAssertError isn't @nogc either.
+    ref inout(T) assumeNonEmpty() inout @property pure @safe
+    {
+        import core.exception : AssertError;
+        import std.exception : enforce;
+        enforce!AssertError(isDefined, T.stringof ~ " assumption failed to hold");
+        return value;
+    }
+
+    /**
+     * If there is a held value, return it; otherwise return other.
+     * Params: other = value to return if this instance is empty
+     * Bugs: These functions are currently not `nothrow` or `@nogc`. See
+     * $(BUGZILLA 12647).
+     * See_Also: $(REF Algebraic, std, variant)
+     */
+    // TODO: Do lazy parameters offer ref yet?
+    auto ref CommonType!(inout(T), U) getOrElse(U)(lazy U other) inout
+        if (!is(CommonType!(T, U) == void))
+    {
+        return isDefined ? value : other;
+    }
+
+    /// Ditto
+    Option!(CommonType!(T, U)) orElse(U)(lazy Option!U other)
+        if (!is(CommonType!(T, U) == void))
+    {
+        alias Super = CommonType!(T, U);
+        return isDefined ? only!Super(this.value) : other.upcast!Super;
+    }
+
+    version(StdDdoc)
+    {
+        /// Ditto
+        Algebraic!(inout(T), U) or(U)(lazy U other) inout;
+    }
+    else // inout does not support field variables at this time
+    {
+        template or(U)
+        {
+            import std.variant : Algebraic;
+            Algebraic!(T, U) or(lazy U other)
+            {
+                return isDefined ? typeof(return)(value) : typeof(return)(other);
+            }
+
+            Algebraic!(const T, U) or(lazy U other) const
+            {
+                return isDefined ? typeof(return)(value) : typeof(return)(other);
+            }
+
+            Algebraic!(immutable T, U) or(lazy U other) immutable
+            {
+                return isDefined ? typeof(return)(value) : typeof(return)(other);
+            }
+        }
+    }
+}
+
+/// Convert from null to Option as soon as possible when interfacing with code that uses null:
+unittest
+{
+    import std.algorithm.comparison : equal;
+    import std.algorithm.iteration : map, joiner;
+    import std.range : only;
+
+    static shared(int)* getX(bool yes)
+    {
+        static shared int x = 42;
+        return yes? &x : null;
+    }
+
+    auto values = only(false, true, false).map!getX.map!option;
+    assert(values[0].empty);
+    assert(values[1].map!(p => *p).equal(only(42)));
```

Comment:
Could add an assert for `values[1].equal(only(getX(true)))` before this line. (And if we add `deref`, we can use it instead of map here and on line 7044).